### DISCUSSION TITLE
Anyone willing and able to make an update to the MPV_alt WLX plugin

### DISCUSSION THREAD

**Discussion by advice2020** [doublecmd/doublecmd#1782] [General]
For anyone unaware and interested, there is a WLX plugin called MPV_alt that is great plugin that allows users to use MPV within DC.  This plugin can be found here
https://github.com/j2969719/doublecmd-plugins/tree/master/plugins/wlx/mpv_alt
I do not think the developer wants to make any more updates to the plugin, so I was wondering if anyone has the ability / interest in making an update to this plugin.

In MPV itself it allows you to use, I believe the RMB by default to click on the main pane to toggle pause/play.  There is also a setting that allows users to choose to also achieve this using the LMB.
Because this WLX plugin can only use mouse to control it, I think this might be a nice addition to the plugin if it is possible.  Sometimes the play/pause icon can get a bit small, this will allow the entire main pane to act as a play/pause toggle button.

Does anyone know how / if this MPV mouse click setting can somehow be accessed in the MPV API used in this MPV_alt WLX plugin.

Hopefully this makes sense.
Thank you to anyone interested

**Comment by j2969719** [comment]
https://mpv.io/manual/master/#lua-scripting-mp-add-forced-key-binding(-)

eg try add to plugload.lua something like
```lua
function on_lmbclick()
  mp.commandv("cycle", "pause")
end
mp.add_forced_key_binding("MBTN_LEFT", on_lmbclick)
```

**Comment by advice2020** [comment]
@j2969719
First off thank you for your response, I apologize I didn't ask you specifically because I was not sure if you were interested in updating this plugin any further.
Second..... what!, this works perfectly!  I really didn't think I was going to get a solution for this one.
Strange how most of the things I think are not going to be possible end up being easier than expected, yet the things I think will be simple end up not being so easy.
Anyways thank you again for your help with this and for all the help / plugins you have provided me in the past.

**Comment by advice2020** [comment]
I have been messing around with similar code that was provided above to see what other mouse based actions I could get to work.  I have figured a few things out but I had a question if anyone has some time.
I do not think this is possible because I do not think that focus can be given to the video pane in this way but
I was wondering it was possible for the mouse scroll wheel to be used on the video pane allowing access to actions assigned to it.
For example in MPV itself if you mouse scroll on the video pane it adjusts the volume up & down.  I was trying to see if I could replicate this volume action or maybe a different action like seek forward/back by 5 and -5.

I tried some things like the following
```
function on_wheelup()
  mp.commandv("cycle", "volume")
end
mp.add_forced_key_binding("MBTN_WHEEL_UP", on_wheelup)
```
but this did not work
if this can work I also do not know how to split the command into two separate actions in this case because volume requires up and down actions
`MBTN_WHEEL_UP`
`MBTN_WHEEL_DOWN`

When messing around hard to know what is not working because not right and what is not possible to achieve in the plugin itself.

Thank you for any help provided

**Reply by j2969719** [reply]
>`MBTN_WHEEL_UP`

`WHEEL_UP`

>`mp.commandv("cycle", "volume")`

`mp.commandv("add", "volume",  -5)`
or
`mp.commandv("add", "volume", 5)`

`mp.commandv("cycle", "volume", -5)`
should also work if thats what you need

**Comment by advice2020** [comment]
@j2969719 
Ahh thank you very much, this works great.
Okay, WHEEL_UP, the manual doesn't even say anything under this item "WHEEL_" I see it a couple options up briefly mentioned.
Hate when I do that, I literally have used those volume actions before for hotkeys before, I was just too focused on the "cycle"
Thank you again.

If anyone is interested here is the seek forward / backward I mentioned
`mp.commandv("seek", -5)`
and
`mp.commandv("seek", 5)`

**Comment by advice2020** [comment]
`USING CUSTOM OSC`
I was looking to be able to possibly use a custom OSC with this plugin.
I read that the following items can be added to achieve this but I do not know where to put them in the `plugin.c` or `plugload.lua` file?
```
mpv_set_option_string(mpv, "config-dir", "/my/path"); // set config root
mpv_set_option_string(mpv, "config", "yes"); // enable config loading
```
Thank You

**Reply by j2969719** [reply]
try adding these `mpv_set_option_string` between lines 28..29 in `plugin.c` https://github.com/j2969719/doublecmd-plugins/blob/b2232367d0337ebff5a37e69be0e26a5829fc4d6/plugins/wlx/mpv_alt/src/plugin.c#L28

by default the profile for embedding in other applications is used + these options may break compatibility with earlier versions so i didnt use them

**Comment by advice2020** [comment]
@j2969719 
Thank you for your response.
At first I originally tried `$HOME/.config/mpv_alt/` & `~/.config/mpv_alt/` for my paths but neither of them worked.  I had to actually type `home/<username>/` in the path to get it to work.  Would rather be able to use either of them rather than include username specific path.  Do you know if this is just required, or if I typed these wrong?

I agree with your decision, if this could cause an issue for other users, better to not include.  They would have to change the path anyways, unless you are able to define one for them?
Does not matter to me because I now know about this, but if interested maybe adding these lines commented out with a note, for other users in the future.  I do not know how many users would be interested in this, so up to you if feel worth adding.
```
//enable the following two lines and update path to allow for custom MPV config
//mpv_set_option_string(mpv, "config-dir", "/my/path"); // set config root
//mpv_set_option_string(mpv, "config", "yes"); // enable config loading
```
Regardless though thank you this works great with the little testing I have done so far.

**Comment by advice2020** [comment]
`INSTRUCTIONS: HOW TO USE CUSTOM OSC`
If any users are interested in the future here are some instructions on how to do so.....
-Create the following folder & subfolder in your .config folder and place your osc.lua file into it
`/mpv_alt/scripts/osc.lua`
-Then in your `plugload.lua` file add the following line which turns off the default MPV OSC
`mp.set_property("osc", "no")`
-Then before compiling add the following two lines containing `mpv_set_option_string` in between lines 28 & 29 in the `plugin.c` file (as and update the path in the first line to the path you used above.
Note:
I originally tried `$HOME/.config/mpv_alt/` & `~/.config/mpv_alt/` neither of them worked, I had to actually put in my actual home/user name to get this to work
`/home/<username>/.config/mpv_alt/`
```
	mpv_handle *mpv = mpv_create();
	mpv_set_option_string(mpv, "config-dir", "/home/<username>/.config/mpv_alt/");
	mpv_set_option_string(mpv, "config", "yes");
	mpv_initialize(mpv); //
```
Then launch DC and it should work.
Hopefully someone will find this useful in the future.
